src/Processors/Formats/Impl/ParallelParsingInputFormat.cpp

```diff
@@ -0,0 +1,224 @@
+#include <Processors/Formats/Impl/ParallelParsingInputFormat.h>
+#include <IO/ReadHelpers.h>
+
+namespace DB
+{
+
+void ParallelParsingInputFormat::segmentatorThreadFunction()
+{
+    setThreadName("Segmentator");
+    try
+    {
+        while (!parsing_finished)
+        {
+            const auto current_unit_number = segmentator_ticket_number % processing_units.size();
+            auto & unit = processing_units[current_unit_number];
+
+            {
+                std::unique_lock<std::mutex> lock(mutex);
+                segmentator_condvar.wait(lock,
+                                         [&]{ return unit.status == READY_TO_INSERT || parsing_finished; });
+            }
+
+            if (parsing_finished)
+            {
+                break;
+            }
+
+            assert(unit.status == READY_TO_INSERT);
+
+            // Segmentating the original input.
+            unit.segment.resize(0);
+
+            const bool have_more_data = file_segmentation_engine(in, unit.segment, min_chunk_bytes);
+
+            unit.is_last = !have_more_data;
+            unit.status = READY_TO_PARSE;
+            scheduleParserThreadForUnitWithNumber(segmentator_ticket_number);
+            ++segmentator_ticket_number;
+
+            if (!have_more_data)
+            {
+                break;
+            }
+        }
+    }
+    catch (...)
+    {
+        onBackgroundException();
+    }
+}
+
+void ParallelParsingInputFormat::parserThreadFunction(size_t current_ticket_number)
+{
+    try
+    {
+        setThreadName("ChunkParser");
+
+        const auto current_unit_number = current_ticket_number % processing_units.size();
+        auto & unit = processing_units[current_unit_number];
+
+        assert(unit.segment.size() > 0);
+
+        /*
+         * This is kind of suspicious -- the input_process_creator contract with
+         * respect to multithreaded use is not clear, but we hope that it is
+         * just a 'normal' factory class that doesn't have any state, and so we
+         * can use it from multiple threads simultaneously.
+         */
+        ReadBuffer read_buffer(unit.segment.data(), unit.segment.size(), 0);
+
+        if (current_ticket_number == 0)
+            prepareReadBuffer(read_buffer);
+
+        InputFormatPtr input_format = internal_parser_creator(read_buffer);
```

Comment:
But it's a little different, I think. We still need to recreate the ReadBuffer and put it in InputFormat, because otherwise freeing the buffer or resetting its state is very difficult due to the inconvenient API.